File: LAB6/LAB6_ejercicio.r
# Algebra Lineal - Lab6
# Entrega: 9/28/2020
# @steven_wilson

require(Matrix)
require(pracma)

#Funcion para generar el espacio nulo
EspacioNulo <- function (A) {
  m <- dim(A)[1]; n <- dim(A)[2]
  QR <- base::qr.default(A)
  r <- QR$rank
  if ((r < min(m, n)) || (m < n)) {
    R <- QR$qr[1:r, , drop = FALSE]
    P <- QR$pivot
    F <- R[, (r + 1):n, drop = FALSE]
    I <- base::diag(1, n - r)
    B <- -1.0 * base::backsolve(R, F, r)
    Y <- base::rbind(B, I)
    X <- Y[base::order(P), , drop = FALSE]
    return(X)
  }
  return(base::matrix(0, n, 1))
}




A <- matrix(c(1,1,  0,1, -1,1),2,3) 
A

B <- matrix(c(1,1,1,  1,5,-1,  -1,1,-2),3,3) 
B

C <- matrix(c(1,0,0,  1,1,1,  0,-1,-1, 1,1,-1),3,4) 
C

D <- matrix(c(2,-1,1, -4,2,-2, 0,1,1, 2,2,4, 1,3,4),3,5) 
D

#Encontrar el rango
RangoA <- rankMatrix(A)[1]
RangoA

RangoB <- rankMatrix(B)[1]
RangoB

RangoC <- rankMatrix(C)[1]
C

RangoD <- rankMatrix(D)[1]
RangoD

#Encontrar la nulidad
nulidadA <- ncol(A)-RangoA
nulidadA

nulidadB <- ncol(B)-RangoA
nulidadB

nulidadC <- ncol(C)-RangoA
nulidadC

nulidadD <- ncol(D)-RangoA
nulidadD

#Encontrar el espacio nulo
EspacioNuloA <- EspacioNulo(A)
EspacioNuloA

EspacioNuloB <- EspacioNulo(B)
EspacioNuloB

EspacioNuloC <- EspacioNulo(C)
EspacioNuloC

EspacioNuloD <- EspacioNulo(D)
EspacioNuloD

File: LAB6/LAB6_ejercicio.py
"""
Algebra Lineal - Lab6
Entrega: 9/28/2020
@steven_wilson
"""

import numpy as np
import sympy as sp
from scipy.linalg import null_space


A = sp.Matrix([[1, 0, -1],
               [1, 1, 1]])

B = sp.Matrix([[1, 1, -1],
               [1, 5, 1],
               [1, -1, -2]])

C = sp.Matrix([[1, 1, 0, 1],
               [0, 1, -1, 1],
               [0, 1, -1, -1]])

D = sp.Matrix([[2, -4, 0, 2, 1],
               [-1, 2, 1, 2, 3],
               [1, -2, 1, 4, 4]])

print(A)
print('\n', B)
print('\n', C)
print('\n', D)

# Encontrar el rango
RangoA = np.linalg.matrix_rank(np.array(A).astype(np.float64))
RangoB = np.linalg.matrix_rank(np.array(B).astype(np.float64))
RangoC = np.linalg.matrix_rank(np.array(C).astype(np.float64))
RangoD = np.linalg.matrix_rank(np.array(D).astype(np.float64))

print("\n El rango de A es: "+str(RangoA))
print("\n El rango de B es: "+str(RangoB))
print("\n El rango de C es: "+str(RangoC))
print("\n El rango de D es: "+str(RangoD))


# Encontrar la nulidad
num_rows, num_cols = np.array(A).astype(np.float64).shape
nulidadA = num_cols-RangoA

num_rows, num_cols = np.array(B).astype(np.float64).shape
nulidadB = num_cols-RangoB

num_rows, num_cols = np.array(C).astype(np.float64).shape
nulidadC = num_cols-RangoC

num_rows, num_cols = np.array(D).astype(np.float64).shape
nulidadD = num_cols-RangoD

print("\n La nulidad de A es: " + str(nulidadA))
print("\n La nulidad de B es: " + str(nulidadB))
print("\n La nulidad de C es: " + str(nulidadC))
print("\n La nulidad de D es: " + str(nulidadD))


# Encontrar el espacio Nulo
EspacioNuloA = A.nullspace()
EspacioNuloB = B.nullspace()
EspacioNuloC = C.nullspace()
EspacioNuloD = D.nullspace()

print("\n El espacio nulo de A es: \n "+str(EspacioNuloA))
print("\n El espacio nulo de B es: \n "+str(EspacioNuloB))
print("\n El espacio nulo de C es: \n "+str(EspacioNuloC))
print("\n El espacio nulo de D es: \n "+str(EspacioNuloD))
